src/refdb_fs.c

```diff
@@ -1375,12 +1422,17 @@ static int refdb_fs_backend__delete_tail(
 
 	git_sortedcache_wunlock(backend->refcache);
 
-	if (error == GIT_ENOTFOUND) {
-		error = loose_deleted ? 0 : ref_error_notfound(ref_name);
+	if (error != GIT_ENOTFOUND && (error = packed_write(backend)) < 0)
+		goto cleanup;
+	else if (error == GIT_ENOTFOUND && !is_loose) {
+		error = ref_error_notfound(ref_name);
 		goto cleanup;
 	}
 
-	error = packed_write(backend);
+	/* unlock here, so that we can delete the all empty parents */
+	git_filebuf_cleanup(file);
+
+	error = refdb_fs_backend__delete_empty_parents(backend, ref_name, git_buf_cstr(&loose_path), false);
```

Comment:
I think we should rename this function. It doesn't only remove parents, but also the reference itself.